mcs/class/System/Mono.Btls/MonoBtlsProvider.cs

```diff
@@ -201,21 +201,80 @@ internal static bool ValidateCertificate (MonoBtlsX509Chain chain, MonoBtlsX509V
 			}
 		}
 
+		internal static void SetupCertificateStore (MonoBtlsX509Store store, MonoTlsSettings settings, bool server)
+		{
+#if MONODROID
+				SetupCertificateStore (store);
+				return;
+#else
+
+			if (settings == null || settings.CertificateSearchPaths == null) {
+				SetupCertificateStore (store);
+				return;
+			}
+
+			foreach (var path in settings.CertificateSearchPaths) {
+				if (string.Equals (path, "@default", StringComparison.Ordinal)) {
+					AddTrustedRoots (store, settings, server);
+					AddUserStore (store);
+					AddMachineStore (store);
+				} else if (string.Equals (path, "@user", StringComparison.Ordinal))
+					AddUserStore (store);
+				else if (string.Equals (path, "@machine", StringComparison.Ordinal))
+					AddMachineStore (store);
+				else if (string.Equals (path, "@trusted", StringComparison.Ordinal))
+					AddTrustedRoots (store, settings, server);
+				else if (path.StartsWith ("@pem:", StringComparison.Ordinal)) {
+					var realPath = path.Substring (5);
+					if (Directory.Exists (realPath))
+						store.AddDirectoryLookup (realPath, MonoBtlsX509FileType.PEM);
+				} else if (path.StartsWith ("@der:", StringComparison.Ordinal)) {
+					var realPath = path.Substring (5);
+					if (Directory.Exists (realPath))
+						store.AddDirectoryLookup (realPath, MonoBtlsX509FileType.ASN1);
+				} else {
+					if (Directory.Exists (path))
```

Comment:
If I understand it correctly, the SetupCertificateStore() method is called for every connection right? It sounds a bit expensive to me to do the Directory.Exists check every time then.

Could we instead move this logic into MonoTlsSettings.cs so the path is checked just once?
And if we're doing that, we could also get rid of the magic strings and do something like:

```csharp
enum CertificateSearchPathType {
  Default,
  User,
  Machine,
  Trusted,
  Pem,
  Der
}

public void AddCertificateSearchPath (CertificateSearchPathType type, string path);

```